File: cs/Graphs/CostNetwork.cs
using System;
using System.Collections.Generic;

namespace Algorithms.Graphs
{
	public class CostNetwork
	{
		private class Edge
		{
			public readonly int Source, Dest, Capacity, Cost;
			public Edge Other;
			public int Flow;

			public Edge(int from, int to, int cost, int capacity)
			{
				Source = from;
				Dest = to;
				Capacity = capacity;
				Cost = cost;
			}
		}

		private int noNodes = 0;
		private readonly List<Edge> edges = new List<Edge>();
		
		public void AddEdge(int source, int dest, int capacity, int cost)
		{
			if (source + 1 > noNodes) 
				noNodes = source + 1;
			if (dest + 1 > noNodes) 
				noNodes = dest + 1;
			Edge e1 = new Edge(source, dest, cost, capacity);
			Edge e2 = new Edge(dest, source, -cost, 0);
			e1.Other = e2;
			e2.Other = e1;
			edges.Add(e1);
			edges.Add(e2);
		}

		private int BellmanFord(int source, out Edge[] p)
		{
			p = new Edge[noNodes];
			int[] d = new int[noNodes];
			for (int i = 0; i < noNodes; i++)
				d[i] = int.MaxValue;
			d[source] = 0;
			bool change = true;
			for (int m = 1; change && m < noNodes - 1; m++)
			{
				change = false;
				foreach(Edge e in edges)
					if (e.Capacity > e.Flow && d[e.Source] != int.MaxValue && d[e.Dest] > d[e.Source] + e.Cost)
					{
						change = true;
						d[e.Dest] = d[e.Source] + e.Cost;
						p[e.Dest] = e;
					}
			}
			foreach(Edge e in edges)
				if (e.Capacity > e.Flow && d[e.Source] != int.MaxValue && d[e.Dest] > d[e.Source] + e.Cost)
				{
					p[e.Dest] = e;
					return e.Dest;
				}
			return -1;
		}

		private static void IncrementFlow(int source, int sink, Edge[] p)
		{
			int minCapacity = int.MaxValue;
			for (int u = sink; u != source; u = p[u].Source)
				minCapacity = Math.Min(minCapacity, p[u].Capacity - p[u].Flow);
			
//			string s = sink.ToString();
			for (int u = sink; u != source; u = p[u].Source)
			{
//				s = p[u].Source + " -> " + s;
				p[u].Flow += minCapacity;
				p[u].Other.Flow -= minCapacity;
			}
//			s += ": Increase flow with " + minCapacity;
//			Console.WriteLine(s);
		}

		/*private static void AdjustCycle(int u, Edge[] p)
		{
//			int dst = p[u].Cost;
			int minCapacity = p[u].Capacity - p[u].Flow;
			for (int v = p[u].Source; v != u; v = p[v].Source)
			{
				minCapacity = Math.Min(minCapacity, p[v].Capacity - p[v].Flow);
//				dst += p[v].Cost;
			}
			p[u].Flow += minCapacity;
			p[u].Other.Flow -= minCapacity;
			string s = p[u].Source.ToString();
			for (int v = p[u].Source; v != u; v = p[v].Source)
			{
				s = p[v].Source + " -> " + s;
				p[v].Flow += minCapacity;
				p[v].Other.Flow -= minCapacity;
			}
			s += ": Adjusting flow with " + minCapacity;
			Console.WriteLine(s);
		}*/

		public int[] MaxFlowMinCost(int source, int sink)
		{
			while (true)
			{
				Edge[] p;
				int u = BellmanFord(source, out p);
				if (u == -1)
				{
					if (p[sink] == null)
						break;
					IncrementFlow(source, sink, p);
				}
				else
					throw new Exception(); // Shouldn't happen
//					AdjustCycle(u, p);
			}

			int flow = 0, cost = 0;
			foreach (Edge edge in edges)
			{
				if (edge.Source == source)
					flow += edge.Flow;

				if (edge.Capacity > 0)
					cost += edge.Cost * edge.Flow;
			}
			return new int[] { flow, cost };
		}
	}
}

File: java/src/yarin/yal/graph/CostNetwork.java
package yarin.yal.graph;

import java.util.ArrayList;
import java.util.List;

// Test at Kattis, https://open.kattis.com/problems/mincostmaxflow
public class CostNetwork {
    private class Edge {
        private final int source, dest, capacity, cost;
        private Edge other;
        private int flow;

        public Edge(int from, int to, int cost, int capacity) {
            this.source = from;
            this.dest = to;
            this.capacity = capacity;
            this.cost = cost;
        }
    }

    private int noNodes = 0;
    private final List<Edge> edges = new ArrayList<Edge>();

    public void addEdge(int source, int dest, int capacity, int cost) {
        if (source + 1 > noNodes) {
            noNodes = source + 1;
        }
        if (dest + 1 > noNodes) {
            noNodes = dest + 1;
        }
        Edge e1 = new Edge(source, dest, cost, capacity);
        Edge e2 = new Edge(dest, source, -cost, 0);
        e1.other = e2;
        e2.other = e1;
        edges.add(e1);
        edges.add(e2);
    }

    private int bellmanFord(int source, Edge[] p) {
        int[] d = new int[noNodes];
        for (int i = 0; i < noNodes; i++) {
            d[i] = Integer.MAX_VALUE;
        }
        d[source] = 0;
        boolean change = true;
        for (int m = 0; change && m < noNodes - 1; m++) {
            change = false;
            for (Edge e : edges) {
                if (e.capacity > e.flow && d[e.source] != Integer.MAX_VALUE && d[e.dest] > d[e.source] + e.cost) {
                    change = true;
                    d[e.dest] = d[e.source] + e.cost;
                    p[e.dest] = e;
                }
            }
        }
        for(Edge e : edges) {
            if (e.capacity > e.flow && d[e.source] != Integer.MAX_VALUE && d[e.dest] > d[e.source] + e.cost) {
                p[e.dest] = e;
                return e.dest;
            }
        }
        return -1;
    }

    private static void incrementFlow(int source, int sink, Edge[] p) {
        int minCapacity = Integer.MAX_VALUE;
        for (int u = sink; u != source; u = p[u].source) {
            minCapacity = Math.min(minCapacity, p[u].capacity - p[u].flow);
        }

//			string s = sink.ToString();
        for (int u = sink; u != source; u = p[u].source) {
//				s = p[u].Source + " -> " + s;
            p[u].flow += minCapacity;
            p[u].other.flow -= minCapacity;
        }
//			s += ": Increase f with " + minCapacity;
//			Console.WriteLine(s);
    }

    public long[] maxFlowMinCost(int source, int sink) {
        while (true) {
            Edge[] p = new Edge[noNodes];;
            int u = bellmanFord(source, p);
            if (u == -1) {
                if (p[sink] == null) {
                    break;
                }
                incrementFlow(source, sink, p);
            }
            else {
                throw new RuntimeException(); // Shouldn't happen
            }
        }

        long flow = 0, cost = 0;
        for (Edge edge : edges) {
            if (edge.source == source) {
                flow += edge.flow;
            }

            if (edge.capacity > 0) {
                cost += (long) edge.cost * edge.flow;
            }
        }
        return new long[] { flow, cost };
    }
}
